gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/proc/DagProcUtils.java

```diff
@@ -291,4 +291,79 @@ public static void removeFlowFinishDeadlineDagAction(DagManagementStateStore dag
       log.warn("Failed to delete dag action {}", enforceFlowFinishDeadlineDagAction);
     }
   }
+
+  /**
+   * Returns true if all dag nodes are finished, and it is not possible to run any new dag node.
+   */
+  public static boolean isDagFinished(Dag<JobExecutionPlan> dag) {
+    List<Dag.DagNode<JobExecutionPlan>> nodes = dag.getNodes();
+    Set<Dag.DagNode<JobExecutionPlan>> canRun = new HashSet<>(nodes);
+    Set<Dag.DagNode<JobExecutionPlan>> completed = new HashSet<>();
+    DagManager.FailureOption failureOption = DagManagerUtils.getFailureOption(dag);
+    boolean anyFailure = false;
+
+    for (Dag.DagNode<JobExecutionPlan> node : nodes) {
+      if (!canRun.contains(node)) {
+        continue;
+      }
+      ExecutionStatus status = node.getValue().getExecutionStatus();
+      if (status == ExecutionStatus.FAILED || status == ExecutionStatus.CANCELLED) {
+        anyFailure = true;
+        removeChildrenFromCanRun(node, dag, canRun);
```

Comment:
is there a specific ordering guarantee for `DagNode<T>::getNodes`?  how do we know this parent would be executed after the children, such that it's children don't get added to `canRun`, even though their parent's failure should mean the child isn't ready to run?

Reply:
hmm, I have not checked the code that generates the Dag, but I think there is a possibility of ordering.
I think this algo would work for any order of dag nodes. A node may not be removed from `canRun` when it is parsed, but it should eventually get removed if it should be.
I added tests to verify this. Randomized the dag nodes and tests pass.

Reply:
good call on those test mods.  they give me far greater confidence.  given that determinism is typically preferable in tests, please do leave a comment to explain what you're verifying w/ the randomness, so future maintainers understand